File: dev/core/src/com/google/gwt/dev/util/SelectionScriptTemplate.java
package com.google.gwt.dev.util;

public class SelectionScriptTemplate {

  public static String answerFunction() {
    return "  // Deferred-binding mapper function.\n" + "  //\n"
        + "  function O(a,v) {\n" + "    var answer = O.answers;\n"
        + "    var i = -1;\n" + "    var n = a.length - 1;\n"
        + "    while (++i < n) {\n" + "      if (!(a[i] in answer)) {\n"
        + "        answer[a[i]] = [];\n" + "      }\n"
        + "      answer = answer[a[i]];\n" + "    }\n"
        + "    answer[a[n]] = v;\n" + "  }\n" + "  O.answers = [];\n" + "\n";
  }

  public static String cssInjector(String cssUrl) {
    return "  if (!__gwt_stylesLoaded['" + cssUrl + "']) {\n"
        + "    __gwt_stylesLoaded['" + cssUrl + "'] = true;\n"
        + "    document.write('<link rel=\\\"stylesheet\\\" href=\\\"" + cssUrl
        + "\\\">');\n" + "  }\n";
  }

  public static String fixedHeader() {
    return "if (!window.__gwt_stylesLoaded) { window.__gwt_stylesLoaded = {}; }\n"
        + "if (!window.__gwt_scriptsLoaded) { window.__gwt_scriptsLoaded = {}; }\n"
        + "\n"

        + "__gwt_getMetaProperty = function(name) {\n"
        + "  var value = __gwt_metaProps[name];\n"
        + "  if (value) {\n"
        + "    return value;\n"
        + "  } else {\n"
        + "    return null;\n"
        + "  }\n"
        + "}\n"
        + "\n"
        +

        "// Called from compiled code to hook the window's resize & load events (the\n"
        + "// code running in the script frame is not allowed to hook these directly).\n"
        + "//\n"
        + "function __gwt_initHandlers(resize, beforeunload, unload) {\n"
        + "  var oldOnResize = window.onresize;\n"
        + "  window.onresize = function() {\n"
        + "    resize();\n"
        + "    if (oldOnResize)\n"
        + "      oldOnResize();\n"
        + "  };\n"
        + "\n"
        + "  var oldOnBeforeUnload = window.onbeforeunload;\n"
        + "  window.onbeforeunload = function() {\n"
        + "    var ret = beforeunload();\n"
        + "\n"
        + "    var oldRet;\n"
        + "    if (oldOnBeforeUnload)\n"
        + "      oldRet = oldOnBeforeUnload();\n"
        + "\n"
        + "    if (ret !== null)\n"
        + "      return ret;\n"
        + "    return oldRet;\n"
        + "  };\n"
        + "\n"
        + "  var oldOnUnload = window.onunload;\n"
        + "  window.onunload = function() {\n"
        + "    unload();\n"
        + "    if (oldOnUnload)\n"
        + "      oldOnUnload();\n"
        + "  };\n"
        + "}\n" + "\n";
  }

  public static String moduleFunctionHeader() {
    return "  // These variables contain deferred-binding properties, values, and\n"
        + "  // providers.\n"
        + "  //\n"
        + "  var props = [];\n"
        + "  var values = [];\n" + "  var providers = [];\n";
  }

  public static String outerFunctions(String moduleName) {
    String moduleFunction = moduleName.replace('.', '_');

    return "// This is global because it is needed by onScriptLoad().\n"
        + "//\n"
        + moduleFunction
        + ".isHostedMode = function() {\n"
        + "  return (window.external && window.external.gwtOnLoad &&\n"
        + "    (document.location.href.indexOf('gwt.hybrid') == -1));\n"
        + "}\n"
        + "\n"
        +

        "// Called by both onScriptLoad() and onInjectionDone(). It causes\n"
        + "// the specified module to be cranked up.\n"
        + "//\n"
        + moduleFunction
        + ".maybeStartModule = function() {\n"
        + "  if ("
        + moduleFunction
        + ".scriptsDone && "
        + moduleFunction
        + ".loadDone) {\n"
        + "    var iframe = document.getElementById('"
        + moduleName
        + "');\n"
        + "    iframe.contentWindow.gwtOnLoad(function() { "
        + moduleFunction
        + ".onBadLoad() },\n"
        + "      '"
        + moduleName
        + "');\n"
        + "  }\n"
        + "}\n"
        + "\n"
        +

        "// Called when the compiled script identified by moduleName is done loading.\n"
        + "// It needs the iframe's window object for starting hosted mode.\n"
        + "//\n"
        + moduleFunction
        + ".onScriptLoad = function(wnd) {\n"
        + "  // If we're running in hosted mode, call window.external.gwtOnLoad().\n"
        + "  if ("
        + moduleFunction
        + ".isHostedMode()) {\n"
        + "    var rc = window.external.gwtOnLoad(wnd, '"
        + moduleName
        + "');\n"
        + "    if (!rc) {\n"
        + "      if (window.__gwt_onLoadError) {\n"
        + "        window.__gwt_onLoadError(name);\n"
        + "      } else {\n"
        + "        alert('Failed to load module "
        + moduleName
        + "' +\n"
        + "          '. Please see the log in the development shell for details.');\n"
        + "      }\n"
        + "    }\n"
        + "    return;\n"
        + "  }\n"
        + "\n"
        + "  // Mark this module's script as done loading and (possibly) start the module.\n"
        + "  "
        + moduleFunction
        + ".loadDone = true;\n"
        + "  "
        + moduleFunction
        + ".maybeStartModule();\n"
        + "}\n"
        + "\n"
        +

        "// Called when the script injection is complete.\n"
        + "//\n"
        + moduleFunction
        + ".onInjectionDone = function() {\n"
        + "  // Mark this module's script injection as done and (possibly) start the\n"
        + "  // module.\n"
        + "  "
        + moduleFunction
        + ".scriptsDone = true;\n"
        + "  "
        + moduleFunction
        + ".maybeStartModule();\n"
        + "}\n"
        + "\n"
        +

        "// Called to slurp up all <meta> tags:\n"
        + "//  gwt:property, gwt:base, gwt:onPropertyErrorFn, gwt:onLoadErrorFn\n"
        + "//\n"
        + moduleFunction
        + ".processMetas = function() {\n"
        + "  if (!!window.__gwt_metaProps) {\n"
        + "    return;\n"
        + "  }\n"
        + "  window.__gwt_metaProps = {};\n"
        + "\n"
        + "  var metas = document.getElementsByTagName('meta');\n"
        + "\n"
        + "  for (var i = 0, n = metas.length; i < n; ++i) {\n"
        + "    var meta = metas[i];\n"
        + "    var name = meta.getAttribute('name');\n"
        + "\n"
        + "    if (name) {\n"
        + "      if (name == 'gwt:property') {\n"
        + "        var content = meta.getAttribute('content');\n"
        + "        if (content) {\n"
        + "          var name = content, value = '';\n"
        + "          var eq = content.indexOf('=');\n"
        + "          if (eq != -1) {\n"
        + "            name = content.substring(0, eq);\n"
        + "            value = content.substring(eq+1);\n"
        + "          }\n"
        + "          __gwt_metaProps[name] = value;\n"
        + "        }\n"
        + "      } else if (name == 'gwt:onPropertyErrorFn') {\n"
        + "        var content = meta.getAttribute('content');\n"
        + "        if (content) {\n"
        + "          try {\n"
        + "            window.__gwt_onPropertyError = eval(content);\n"
        + "          } catch (e) {\n"
        + "            alert('Bad handler \\\"' + content +\n"
        + "              '\\\" for \\\"gwt:onPropertyErrorFn\\\"');\n"
        + "          }\n"
        + "        }\n"
        + "      } else if (name == 'gwt:onLoadErrorFn') {\n"
        + "        var content = meta.getAttribute('content');\n"
        + "        if (content) {\n"
        + "          try {\n"
        + "            window.__gwt_onLoadError = eval(content);\n"
        + "          } catch (e) {\n"
        + "            alert('Bad handler \\\"' + content + '\\\" for \\\"gwt:onLoadErrorFn\\\"');\n"
        + "          }\n"
        + "        }\n"
        + "      } else if (name == 'gwt:base') {\n"
        + "        var content = meta.getAttribute('content');\n"
        + "        var eqPos = content.lastIndexOf('=');\n"
        + "        if (eqPos == -1) {\n"
        + "          continue;\n"
        + "        }\n"
        + "        var moduleBase = content.substring(0, eqPos);\n"
        + "        var moduleName = content.substring(eqPos + 1);\n"
        + "        if (!window.__gwt_base) {\n"
        + "          window.__gwt_base = [];\n"
        + "        }\n"
        + "        window.__gwt_base[moduleName] = moduleBase;\n"
        + "      }\n"
        + "    }\n"
        + "  }\n"
        + "}\n"
        + "\n"
        +

        "// Called whenever a bad property is provided.\n"
        + "//\n"
        + moduleFunction
        + ".onBadProperty = function(propName, allowedValues, badValue) {\n"
        + "  if (window.__gwt_onPropertyError) {\n"
        + "    "
        + moduleFunction
        + ".onPropertyError(propName, allowedValues, badValue);\n"
        + "    return;\n"
        + "  } else {\n"
        + "    var msg = 'While attempting to load module "
        + moduleName
        + ", ';\n"
        + "    if (badValue != null) {\n"
        + "      msg += 'property \\\"' + propName + '\\\" was set to the unexpected value \\\"'\n"
        + "        + badValue + '\\\"';\n"
        + "    } else {\n"
        + "      msg += 'property \\\"' + propName + '\\\" was not specified';\n"
        + "    }\n"
        + "\n"
        + "    msg += 'Allowed values: ' + allowedValues;\n"
        + "\n"
        + "    alert(msg);\n"
        + "  }\n"
        + "}\n"
        + "\n"
        +

        "// Called if the module fails to load.\n"
        + "//\n"
        + moduleFunction
        + ".onBadLoad = function() {\n"
        + "  if (window.__gwt_onLoadError) {\n"
        + "    window.__gwt_onLoadError('"
        + moduleName
        + "');\n"
        + "    return;\n"
        + "  } else {\n"
        + "    alert('Failed to load module "
        + moduleName
        + "' +\n"
        + "      '\".\\nPlease see the log in the development shell for details.');\n"
        + "  }\n" + "}\n" + "\n";
  }

  public static String scriptInjector(String scriptUrl) {
    return "  if (!__gwt_scriptsLoaded['" + scriptUrl + "']) {\n"
        + "    __gwt_scriptsLoaded['" + scriptUrl + "'] = true;\n"
        + "    document.write('<script language=\\\"javascript\\\" src=\\\""
        + scriptUrl + "\\\"></script>');\n" + "  }\n";
  }
}

File: dev/core/src/com/google/gwt/dev/util/SelectionScriptTemplate.js
/*
 * Copyright 2006 Google Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
if (!window.__gwt_stylesLoaded) { window.__gwt_stylesLoaded = {}; }
if (!window.__gwt_scriptsLoaded) { window.__gwt_scriptsLoaded = {}; }

// Called from compiled code to hook the window's resize & load events (the
// code running in the script frame is not allowed to hook these directly).
//
function __gwt_initHandlers(resize, beforeunload, unload) {
  var oldOnResize = window.onresize;
  window.onresize = function() {
    resize();
    if (oldOnResize)
      oldOnResize();
  };

  var oldOnBeforeUnload = window.onbeforeunload;
  window.onbeforeunload = function() {
    var ret = beforeunload();

    var oldRet;
    if (oldOnBeforeUnload)
      oldRet = oldOnBeforeUnload();

    if (ret !== null)
      return ret;
    return oldRet;
  };

  var oldOnUnload = window.onunload;
  window.onunload = function() {
    unload();
    if (oldOnUnload)
      oldOnUnload();
  };
}

function __MODULE_FUNC__() {
  // These variables contain deferred-binding properties, values, and
  // providers.
  //
  var props = [];
  var values = [];
  var providers = [];

// __SHELL_SERVLET_ONLY_BEGIN__
  // Forces shell servlet to serve compiled output for web mode
  if (!__MODULE_FUNC__.isHostedMode()) {
    document.write('<script src="__MODULE_NAME__.nocache.js?compiled"></script>');
    return;
  }
// __SHELL_SERVLET_ONLY_END__

// __PROPERTIES_BEGIN__
  // Property providers and values
// __PROPERTIES_END__

  __MODULE_FUNC__.processMetas();

  var strongName;
  if (__MODULE_FUNC__.isHostedMode()) {
    // In hosted mode, inject the script frame directly.
    var iframe = document.createElement('iframe');
    document.body.appendChild(iframe);
    iframe.style.width = iframe.style.height = iframe.style.border = '0px';
    var wnd = iframe.contentWindow;
    wnd.$wnd = window;
    wnd.$doc = window.document;
    wnd.__gwt_getProperty = function(name) { return providers[name](); };
    wnd.onunload = function() { alert('unload!'); };
    __MODULE_FUNC__.onScriptLoad(wnd);
  } else {
    // Deferred-binding mapper function.
    //
    function O(a,v) {
      var answer = O.answers;
      var i = -1;
      var n = a.length - 1;
      while (++i < n) {
        if (!(a[i] in answer)) {
          answer[a[i]] = [];
        }
        answer = answer[a[i]];
      }
      answer[a[n]] = v;
    }
    O.answers = [];

    try {
// __PERMUTATIONS_BEGIN__
      // Permutation logic
// __PERMUTATIONS_END__
    } catch (e) {
      // intentionally silent on property failure
      return;
    }  

	  // TODO: do we still need this query stuff?
	  var query = location.search;
	  query = query.substring(0, query.indexOf('&'));

	  var base;
	  if (window.__gwt_base) {
	    base = __gwt_base['__MODULE_NAME__'];
	  }
	  var newUrl = (base ? base + '/' : '') + strongName + '.cache.html' + query;
	  document.write('<iframe id="__MODULE_NAME__" style="width:0;height:0;border:0" src="' + newUrl + '"></iframe>');
  }

// __MODULE_DEPS_BEGIN__
  // Module dependencies, such as scripts and css
// __MODULE_DEPS_END__
  document.write('<script>__MODULE_FUNC__.onInjectionDone(\'__MODULE_NAME__\')</script>');
}

__MODULE_FUNC__.getMetaProperty = function(name) {
  var value = __MODULE_FUNC__.metaProps[name];
  if (value) {
    return value;
  } else {
    return null;
  }
}

// This is global because it is needed by onScriptLoad().
//
__MODULE_FUNC__.isHostedMode = function() {
  return (window.external && window.external.gwtOnLoad &&
    (document.location.href.indexOf('gwt.hybrid') == -1));
}

// Called by both onScriptLoad() and onInjectionDone(). It causes
// the specified module to be cranked up.
//
__MODULE_FUNC__.maybeStartModule = function() {
  if (__MODULE_FUNC__.scriptsDone && __MODULE_FUNC__.loadDone) {
    var iframe = document.getElementById('__MODULE_NAME__');
    iframe.contentWindow.gwtOnLoad(function() { __MODULE_FUNC__.onBadLoad() },
      '__MODULE_NAME__');
  }
}

// Called when the compiled script identified by moduleName is done loading.
// It needs the iframe's window object for starting hosted mode.
//
__MODULE_FUNC__.onScriptLoad = function(wnd) {
  // If we're running in hosted mode, call window.external.gwtOnLoad().
  if (__MODULE_FUNC__.isHostedMode()) {
    var rc = window.external.gwtOnLoad(wnd, '__MODULE_NAME__');
    if (!rc) {
      if (window.__gwt_onLoadError) {
        window.__gwt_onLoadError(name);
      } else {
        alert('Failed to load module __MODULE_NAME__' +
          '. Please see the log in the development shell for details.');
      }
    }
    return;
  }

  // Mark this module's script as done loading and (possibly) start the module.
  __MODULE_FUNC__.loadDone = true;
  __MODULE_FUNC__.maybeStartModule();
}

// Called when the script injection is complete.
//
__MODULE_FUNC__.onInjectionDone = function() {
  // Mark this module's script injection as done and (possibly) start the
  // module.
  __MODULE_FUNC__.scriptsDone = true;
  __MODULE_FUNC__.maybeStartModule();
}

// Called to slurp up all <meta> tags:
// gwt:property, gwt:base, gwt:onPropertyErrorFn, gwt:onLoadErrorFn
//
__MODULE_FUNC__.processMetas = function() {
  __MODULE_FUNC__.metaProps = {};

  var metas = document.getElementsByTagName('meta');

  for (var i = 0, n = metas.length; i < n; ++i) {
    var meta = metas[i];
    var name = meta.getAttribute('name');

    if (name) {
      if (name == 'gwt:property') {
        var content = meta.getAttribute('content');
        if (content) {
          var name = content, value = '';
          var eq = content.indexOf('=');
          if (eq != -1) {
            name = content.substring(0, eq);
            value = content.substring(eq+1);
          }
          __MODULE_FUNC__.metaProps[name] = value;
        }
      } else if (name == 'gwt:onPropertyErrorFn') {
        var content = meta.getAttribute('content');
        if (content) {
          try {
            window.__gwt_onPropertyError = eval(content);
          } catch (e) {
            alert('Bad handler \"' + content +
              '\" for \"gwt:onPropertyErrorFn\"');
          }
        }
      } else if (name == 'gwt:onLoadErrorFn') {
        var content = meta.getAttribute('content');
        if (content) {
          try {
            window.__gwt_onLoadError = eval(content);
          } catch (e) {
            alert('Bad handler \"' + content + '\" for \"gwt:onLoadErrorFn\"');
          }
        }
      } else if (name == 'gwt:base') {
        var content = meta.getAttribute('content');
        var eqPos = content.lastIndexOf('=');
        if (eqPos == -1) {
          continue;
        }
        var moduleBase = content.substring(0, eqPos);
        var moduleName = content.substring(eqPos + 1);
        if (!window.__gwt_base) {
          window.__gwt_base = [];
        }
        window.__gwt_base[moduleName] = moduleBase;
      }
    }
  }
}

// Called whenever a bad property is provided.
//
__MODULE_FUNC__.onBadProperty = function(propName, allowedValues, badValue) {
  if (window.__gwt_onPropertyError) {
    __MODULE_FUNC__.onPropertyError(propName, allowedValues, badValue);
    return;
  } else {
    var msg = 'While attempting to load module __MODULE_NAME__, ';
    if (badValue != null) {
      msg += 'property \"' + propName + '\" was set to the unexpected value \"'
        + badValue + '\"';
    } else {
      msg += 'property \"' + propName + '\" was not specified';
    }

    msg += 'Allowed values: ' + allowedValues;

    alert(msg);
  }
}

// Called if the module fails to load.
//
__MODULE_FUNC__.onBadLoad = function() {
  if (window.__gwt_onLoadError) {
    window.__gwt_onLoadError('__MODULE_NAME__');
    return;
  } else {
    alert('Failed to load module __MODULE_NAME__' +
      '".\nPlease see the log in the development shell for details.');
  }
}

__MODULE_FUNC__();
